File: assign4/python/src/gameoflife.py
from functools import reduce
from collections import Counter

def is_alive(alive, living_neighbors):
  return alive and living_neighbors == 2 or living_neighbors == 3

def generate_signals(live_cells):
 return sum(
   [[(cell[0] - 1, cell[1] - 1), (cell[0] - 1, cell[1]),
  (cell[0] - 1, cell[1] + 1), (cell[0], cell[1] - 1),
  (cell[0], cell[1] + 1), (cell[0] + 1, cell[1] - 1),
  (cell[0] + 1, cell[1]), (cell[0] + 1, cell[1] + 1)] 
    for cell in live_cells],
  [])

def count_signals_for_each_cell(signals):
  return Counter(signals)
  
def next_generation(live_cells):
  signals = generate_signals(live_cells)

  return [cell for cell, count in count_signals_for_each_cell(signals).items()
    if is_alive(cell in live_cells, count)]

File: assign4/haskell/src/GameOfLife.hs
module GameOfLife where
  import Data.List (groupBy, sort)

  isAlive alive livingNeighbors =
    or [livingNeighbors == 3, alive && livingNeighbors == 2]

  generateSignals liveCells = 
    concat (map (\cell -> [
        ((fst cell) - 1, (snd cell) - 1), ((fst cell) - 1, (snd cell)),
        ((fst cell) - 1, (snd cell) + 1), ((fst cell), (snd cell) - 1),
        ((fst cell), (snd cell) + 1), ((fst cell) + 1, (snd cell) - 1),
        ((fst cell) + 1, (snd cell)), ((fst cell) + 1, (snd cell) + 1)
      ]) liveCells)

  countSignalsForEachCell :: (Ord a, Ord b, Num a, Num b) =>
    [(a, b)] -> [((a, b), Int)]
  countSignalsForEachCell signals =
    map (\listEqual -> (
      head listEqual, length listEqual)) (groupBy (==) (sort signals))

  nextGeneration liveCells =
    let signals = generateSignals liveCells in
    map (\x -> fst x) (filter (\x -> isAlive (elem (fst x) liveCells ) $ snd x) $ countSignalsForEachCell signals)
